File: leetcode/993.cpp
#include <vector>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
private:
  TreeNode *xparent = nullptr;
  TreeNode *yparent = nullptr;
  int xdepth = 0;
  int ydepth = 0;

  void dfs(TreeNode *root, int x, int y, TreeNode *parent, int k) {
    if (root == nullptr) {
      return;
    }
    if (root->val == x) {
      xparent = parent;
      xdepth = k;
      return;
    } else if (root->val == y) {
      yparent = parent;
      ydepth = k;
      return;
    }
    dfs(root->left, x, y, root, k + 1);
    dfs(root->right, x, y, root, k + 1);
  }

public:
  bool isCousins(TreeNode *root, int x, int y) {
    dfs(root, x, y, nullptr, 0);
    return xparent != yparent && xdepth == ydepth;
  }
};

File: leetcode/993.py
from typing import Tuple


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isCousins(self, root: TreeNode, x: int, y: int) -> bool:
        def dfs(root: TreeNode, x: int, y: int, parent: TreeNode, depth: int):
            nonlocal xdepth, ydepth, xparent, yparent
            if not root:
                return

            if root.val == x:
                xparent = parent
                xdepth = depth
                return
            elif root.val == y:
                yparent = parent
                ydepth = depth
                return

            dfs(root.left, x, y, root, depth+1)
            dfs(root.right, x, y, root, depth+1)

        xdepth, ydepth, xparent, yparent = 0, 0, None, None
        dfs(root, x, y, None, 0)
        return xdepth == ydepth and xparent != yparent
